### DISCUSSION TITLE
Any Chance to re-support Tracker.jl?

### DISCUSSION THREAD

**Discussion by patrickm663** [TuringLang/Turing.jl#2432] [General]
Hi all

Could support for something like Tracker.jl be re-introduced, like at a use-at-your-own-risk? If not, is there then a way to manually add in an AD library that isn't explicitly supported ie like Enzyme.jl? Reason being, I have BNNs I've been working on and Tracker.jl was working well enough, and was the fasted backend for the size of the network and data. Updating to 0.35 obviously broke some code, but I couldn't get Mooncake.jl to work as a drop-in replacement sampler.

I then tried replicating the [Bayesian Neural Network tutorial](https://turinglang.org/docs/tutorials/03-bayesian-neural-network/) directly in a clean environment but couldn't get Mooncake to work either. I'm away from my laptop at the moment, but I recall I would get a Stackoverflow error when running the sampler. Other AD libraries worked without modification to the code (although Zygote.jl required that I remove the `f32` conversion). I haven't re-tested this, so it may be fine now :) I also tried loading `AutoTracker()` directly but couldn't get the sampler to work.

Sticking with the tutorial, I downgraded and tested Tracker.jl on the model. It benchmarked slightly behind ForwardDiff.jl (the NN only has ~20 parameters), but beat out Zygote.jl and ReverseDiff.jl. As soon as I upped the size of the NN (200+ parameters), Tracker.jl was a clear winner.

I do get Tracker.jl's not being actively maintained so I agree there shouldn't be tutorials on it, but it's not outright broken yet from my experience.

I would like to benchmark Mooncake.jl soon and see how it compares. I'd also be interested in seeing how performance is on the tutorial docs, since the output is currently suppressed. 

Happy to share a MWE and the stacktrace I get when I'm back at my desk :)

**Comment by torfjelde** [comment]
Mooncake.jl is currently quite slow on Turing.jl due to a bug, so that performance should hopefully be much better very soon:) Ref: #2418 

Regarding Tracker.jl, you can always post the MWE + stacktrace, but I do think it's unlikely that it'll be fixed unless it's obvious what's wrong + a very simple fix :confused:

**Reply by patrickm663** [reply]
Thanks for the quick reply! Tracker's issue is I think just due to `AutoTracker()` getting dropped in the latest release from what I can tell, since it works in v0.34 then breaks in v0.35.

I'll have another look at Mooncake.jl to see if I can get it working for my code, then maybe open a separate Issue if it still breaks. Thanks for the note re the speed, I'll keep an eye on updates to the package :)

**Reply by willtebbutt** [reply]
If you want to attempt to use `Tracker`, you can always import `ADTypes` and use `ADTypes.AutoTracker` directly -- it should work :)

**Reply by willtebbutt** [reply]
Also, re stack overflow with Mooncake -- there's a known issue that I'm actively working on at the minute. It should be resolved tomorrow. Apologies for it not working out-of-the-box!

**Reply by patrickm663** [reply]
Thanks very much for the replies! I have been sticking to Turing v0.34.1 for the time being but will revist getting my set-up to work on the latest version of Turing now that I have some time over the holidays :) In my experience, loading ADTypes let me access `AutoTracker()` but I would get a long stacktrace of `copyto!!()` errors iirc when I then ran the sampler.

**Reply by willtebbutt** [reply]
Ooo that's annoying. I _suspect_ that this is something to do with DifferentiationInterface (I believe that function is defined there), but you'd have to take another look at your stack trace and confirm.

**Reply by gdalle** [reply]
Can you provide an MWE?

**Reply by patrickm663** [reply]
Sure, here is a MWE @gdalle :

```
using Turing, ADTypes, Distributions, StatsPlots, Random, Tracker

@model function foo(y)
	τ ~ truncated(Normal(0, 1); lower=1e-5)
	return y ~ Normal(0, τ^2)
end

y = rand(Xoshiro(1), Normal(0, 2^2), 100);

ad = ADTypes.AutoTracker()

ch_tracker = sample(Xoshiro(2), foo(y), NUTS(; adtype=ad), 500)
```

Error:
```
MethodError: no method matching copyto!(::Float64, ::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{0}, Tuple{}, typeof(+), Tuple{Float64, Float64}})

The function `copyto!` exists, but no method is defined for this combination of argument types.


Closest candidates are:

  copyto!(::RecursiveArrayTools.AbstractVectorOfArray, ::Base.Broadcast.Broadcasted{<:Base.Broadcast.DefaultArrayStyle})

   @ RecursiveArrayTools ~/.julia/packages/RecursiveArrayTools/EU36Y/src/vector_of_array.jl:881

  copyto!(::SciMLBase.DiffEqArrayOperator, ::Any)

   @ SciMLBase ~/.julia/packages/SciMLBase/YTOjh/src/operators/basic_operators.jl:127

  copyto!(::SciMLOperators.MatrixOperator, ::Any)

   @ SciMLOperators ~/.julia/packages/SciMLOperators/KVzmP/src/matrix.jl:185

  ...
```

Full Stack Trace:
<details>

```
Stacktrace:
   [1] materialize!
     @ ./broadcast.jl:883 [inlined]
   [2] materialize!
     @ ./broadcast.jl:880 [inlined]
   [3] (::Tracker.var"#510#512"{TrackedVector{Float64, Vector{Float64}}, Tuple{}})(Δ::Tracker.TrackedReal{Float64})
     @ Tracker ~/.julia/packages/Tracker/6rnwO/src/lib/array.jl:111
   [4] back_(g::Tracker.Grads, c::Tracker.Call{Tracker.var"#510#512"{TrackedVector{Float64, Vector{Float64}}, Tuple{}}, Tuple{Tracker.Tracked{Vector{Float64}}}}, Δ::Tracker.TrackedReal{Float64})
     @ Tracker ~/.julia/packages/Tracker/6rnwO/src/back.jl:110
   [5] back(g::Tracker.Grads, x::Tracker.Tracked{Float64}, Δ::Tracker.TrackedReal{Float64})
     @ Tracker ~/.julia/packages/Tracker/6rnwO/src/back.jl:125
   [6] #710
     @ ~/.julia/packages/Tracker/6rnwO/src/back.jl:113 [inlined]
   [7] #72
     @ ./tuple.jl:693 [inlined]
   [8] BottomRF
     @ ./reduce.jl:86 [inlined]
   [9] _foldl_impl
     @ ./reduce.jl:58 [inlined]
  [10] foldl_impl
     @ ./reduce.jl:48 [inlined]
  [11] mapfoldl_impl
     @ ./reduce.jl:44 [inlined]
  [12] mapfoldl
     @ ./reduce.jl:175 [inlined]
  [13] foldl
     @ ./reduce.jl:198 [inlined]
  [14] foreach
     @ ./tuple.jl:693 [inlined]
  [15] back_(g::Tracker.Grads, c::Tracker.Call{Tracker.var"#266#267"{Tracker.TrackedReal{Float64}}, Tuple{Tracker.Tracked{Float64}}}, Δ::Tracker.TrackedReal{Float64})
     @ Tracker ~/.julia/packages/Tracker/6rnwO/src/back.jl:113
  [16] back(g::Tracker.Grads, x::Tracker.Tracked{Float64}, Δ::Tracker.TrackedReal{Float64})
     @ Tracker ~/.julia/packages/Tracker/6rnwO/src/back.jl:125
  [17] #710
     @ ~/.julia/packages/Tracker/6rnwO/src/back.jl:113 [inlined]
  [18] #72
     @ ./tuple.jl:693 [inlined]
  [19] BottomRF
     @ ./reduce.jl:86 [inlined]
  [20] _foldl_impl(op::Base.BottomRF{Base.var"#72#73"{Tracker.var"#710#711"{Tracker.Grads}}}, init::Nothing, itr::Base.Iterators.Zip{Tuple{Tuple{Tracker.Tracked{Float64}, Nothing}, Tuple{Tracker.TrackedReal{Float64}, Float64}}})
     @ Base ./reduce.jl:58
  [21] foldl_impl
     @ ./reduce.jl:48 [inlined]
  [22] mapfoldl_impl
     @ ./reduce.jl:44 [inlined]
  [23] mapfoldl
     @ ./reduce.jl:175 [inlined]
  [24] foldl
     @ ./reduce.jl:198 [inlined]
  [25] foreach
     @ ./tuple.jl:693 [inlined]
  [26] back_(g::Tracker.Grads, c::Tracker.Call{Tracker.var"#195#198"{Tracker.TrackedReal{Float64}, Float64}, Tuple{Tracker.Tracked{Float64}, Nothing}}, Δ::Tracker.TrackedReal{Float64})
     @ Tracker ~/.julia/packages/Tracker/6rnwO/src/back.jl:113
  [27] back(g::Tracker.Grads, x::Tracker.Tracked{Float64}, Δ::Tracker.TrackedReal{Float64})
     @ Tracker ~/.julia/packages/Tracker/6rnwO/src/back.jl:123
--- the above 11 lines are repeated 1 more time ---
  [39] #710
     @ ~/.julia/packages/Tracker/6rnwO/src/back.jl:113 [inlined]
  [40] #72
     @ ./tuple.jl:693 [inlined]
  [41] BottomRF
     @ ./reduce.jl:86 [inlined]
  [42] _foldl_impl
     @ ./reduce.jl:58 [inlined]
  [43] foldl_impl
     @ ./reduce.jl:48 [inlined]
  [44] mapfoldl_impl
     @ ./reduce.jl:44 [inlined]
  [45] mapfoldl
     @ ./reduce.jl:175 [inlined]
  [46] foldl
     @ ./reduce.jl:198 [inlined]
  [47] foreach
     @ ./tuple.jl:693 [inlined]
  [48] back_(g::Tracker.Grads, c::Tracker.Call{Tracker.var"#406#407"{Tracker.TrackedReal{Float64}}, Tuple{Tracker.Tracked{Float64}}}, Δ::Tracker.TrackedReal{Float64})
     @ Tracker ~/.julia/packages/Tracker/6rnwO/src/back.jl:113
  [49] back(g::Tracker.Grads, x::Tracker.Tracked{Float64}, Δ::Tracker.TrackedReal{Float64})
     @ Tracker ~/.julia/packages/Tracker/6rnwO/src/back.jl:125
  [50] #710
     @ ~/.julia/packages/Tracker/6rnwO/src/back.jl:113 [inlined]
  [51] #72
     @ ./tuple.jl:693 [inlined]
  [52] BottomRF
     @ ./reduce.jl:86 [inlined]
  [53] _foldl_impl
     @ ./reduce.jl:62 [inlined]
--- the above 11 lines are repeated 2 more times ---
  [76] foldl_impl
     @ ./reduce.jl:48 [inlined]
  [77] mapfoldl_impl
     @ ./reduce.jl:44 [inlined]
  [78] mapfoldl
     @ ./reduce.jl:175 [inlined]
  [79] foldl
     @ ./reduce.jl:198 [inlined]
  [80] foreach
     @ ./tuple.jl:693 [inlined]
  [81] back_(g::Tracker.Grads, c::Tracker.Call{Tracker.var"#194#197"{Tracker.TrackedReal{Float64}, Tracker.TrackedReal{Float64}}, Tuple{Tracker.Tracked{Float64}, Tracker.Tracked{Float64}}}, Δ::Tracker.TrackedReal{Float64})
     @ Tracker ~/.julia/packages/Tracker/6rnwO/src/back.jl:113
  [82] back(g::Tracker.Grads, x::Tracker.Tracked{Float64}, Δ::Tracker.TrackedReal{Float64})
     @ Tracker ~/.julia/packages/Tracker/6rnwO/src/back.jl:125
  [83] #710
     @ ~/.julia/packages/Tracker/6rnwO/src/back.jl:113 [inlined]
  [84] #72
     @ ./tuple.jl:693 [inlined]
  [85] BottomRF
     @ ./reduce.jl:86 [inlined]
  [86] _foldl_impl
     @ ./reduce.jl:58 [inlined]
--- the above 11 lines are repeated 10 more times ---
 [197] foldl_impl
     @ ./reduce.jl:48 [inlined]
 [198] mapfoldl_impl
     @ ./reduce.jl:44 [inlined]
 [199] mapfoldl
     @ ./reduce.jl:175 [inlined]
 [200] foldl
     @ ./reduce.jl:198 [inlined]
 [201] foreach
     @ ./tuple.jl:693 [inlined]
 [202] back_(g::Tracker.Grads, c::Tracker.Call{Tracker.var"#194#197"{Tracker.TrackedReal{Float64}, Tracker.TrackedReal{Float64}}, Tuple{Tracker.Tracked{Float64}, Tracker.Tracked{Float64}}}, Δ::Tracker.TrackedReal{Float64})
     @ Tracker ~/.julia/packages/Tracker/6rnwO/src/back.jl:113
 [203] back(g::Tracker.Grads, x::Tracker.Tracked{Float64}, Δ::Tracker.TrackedReal{Float64})
     @ Tracker ~/.julia/packages/Tracker/6rnwO/src/back.jl:125
 [204] #710
     @ ~/.julia/packages/Tracker/6rnwO/src/back.jl:113 [inlined]
 [205] #72
     @ ./tuple.jl:693 [inlined]
 [206] BottomRF
     @ ./reduce.jl:86 [inlined]
 [207] _foldl_impl
     @ ./reduce.jl:62 [inlined]
 [208] foldl_impl
     @ ./reduce.jl:48 [inlined]
 [209] mapfoldl_impl
     @ ./reduce.jl:44 [inlined]
 [210] mapfoldl
     @ ./reduce.jl:175 [inlined]
 [211] foldl
     @ ./reduce.jl:198 [inlined]
 [212] foreach
     @ ./tuple.jl:693 [inlined]
 [213] back_(g::Tracker.Grads, c::Tracker.Call{Tracker.var"#194#197"{Tracker.TrackedReal{Float64}, Tracker.TrackedReal{Float64}}, Tuple{Tracker.Tracked{Float64}, Tracker.Tracked{Float64}}}, Δ::Int64)
     @ Tracker ~/.julia/packages/Tracker/6rnwO/src/back.jl:113
 [214] back(g::Tracker.Grads, x::Tracker.Tracked{Float64}, Δ::Int64)
     @ Tracker ~/.julia/packages/Tracker/6rnwO/src/back.jl:125
 [215] #712
     @ ~/.julia/packages/Tracker/6rnwO/src/back.jl:140 [inlined]
 [216] #715
     @ ~/.julia/packages/Tracker/6rnwO/src/back.jl:149 [inlined]
 [217] logdensity_and_gradient(∇ℓ::LogDensityProblemsADTrackerExt.TrackerGradientLogDensity{LogDensityFunction{DynamicPPL.TypedVarInfo{…}, DynamicPPL.Model{…}, DynamicPPL.SamplingContext{…}}}, x::Vector{Float64})
     @ LogDensityProblemsADTrackerExt ~/.julia/packages/LogDensityProblemsAD/9uEsH/ext/LogDensityProblemsADTrackerExt.jl:41
 [218] ∂logπ∂θ
     @ ~/.julia/packages/Turing/oFGEb/src/mcmc/hmc.jl:176 [inlined]
 [219] ∂H∂θ
     @ ~/.julia/packages/AdvancedHMC/9NhHg/src/hamiltonian.jl:38 [inlined]
 [220] phasepoint
     @ ~/.julia/packages/AdvancedHMC/9NhHg/src/hamiltonian.jl:80 [inlined]
 [221] phasepoint(rng::Xoshiro, θ::Vector{Float64}, h::AdvancedHMC.Hamiltonian{AdvancedHMC.DiagEuclideanMetric{…}, AdvancedHMC.GaussianKinetic, Base.Fix1{…}, Turing.Inference.var"#∂logπ∂θ#36"{…}})
     @ AdvancedHMC ~/.julia/packages/AdvancedHMC/9NhHg/src/hamiltonian.jl:161
 [222] initialstep(rng::Xoshiro, model::DynamicPPL.Model{…}, spl::DynamicPPL.Sampler{…}, vi_original::DynamicPPL.TypedVarInfo{…}; initial_params::Nothing, nadapts::Int64, kwargs::@Kwargs{})
     @ Turing.Inference ~/.julia/packages/Turing/oFGEb/src/mcmc/hmc.jl:180
 [223] step(rng::Xoshiro, model::DynamicPPL.Model{typeof(foo), (:y,), (), (), Tuple{…}, Tuple{}, DynamicPPL.DefaultContext}, spl::DynamicPPL.Sampler{NUTS{…}}; initial_params::Nothing, kwargs::@Kwargs{nadapts::Int64})
     @ DynamicPPL ~/.julia/packages/DynamicPPL/senfM/src/sampler.jl:130
 [224] step
     @ ~/.julia/packages/DynamicPPL/senfM/src/sampler.jl:113 [inlined]
 [225] macro expansion
     @ ~/.julia/packages/AbstractMCMC/FSyVk/src/sample.jl:159 [inlined]
 [226] macro expansion
     @ ~/.julia/packages/ProgressLogging/6KXlp/src/ProgressLogging.jl:328 [inlined]
 [227] (::AbstractMCMC.var"#24#25"{Bool, String, Nothing, Int64, Int64, Int64, Nothing, @Kwargs{…}, Xoshiro, DynamicPPL.Model{…}, DynamicPPL.Sampler{…}, Int64, Int64, Int64})()
     @ AbstractMCMC ~/.julia/packages/AbstractMCMC/FSyVk/src/logging.jl:12
 [228] with_logstate(f::AbstractMCMC.var"#24#25"{Bool, String, Nothing, Int64, Int64, Int64, Nothing, @Kwargs{…}, Xoshiro, DynamicPPL.Model{…}, DynamicPPL.Sampler{…}, Int64, Int64, Int64}, logstate::Base.CoreLogging.LogState)
     @ Base.CoreLogging ./logging/logging.jl:522
 [229] with_logger(f::Function, logger::LoggingExtras.TeeLogger{Tuple{LoggingExtras.EarlyFilteredLogger{…}, LoggingExtras.EarlyFilteredLogger{…}}})
     @ Base.CoreLogging ./logging/logging.jl:632
 [230] with_progresslogger(f::Function, _module::Module, logger::Base.CoreLogging.ConsoleLogger)
     @ AbstractMCMC ~/.julia/packages/AbstractMCMC/FSyVk/src/logging.jl:36
 [231] macro expansion
     @ ~/.julia/packages/AbstractMCMC/FSyVk/src/logging.jl:11 [inlined]
 [232] mcmcsample(rng::Xoshiro, model::DynamicPPL.Model{…}, sampler::DynamicPPL.Sampler{…}, N::Int64; progress::Bool, progressname::String, callback::Nothing, num_warmup::Int64, discard_initial::Int64, thinning::Int64, chain_type::Type, initial_state::Nothing, kwargs::@Kwargs{…})
     @ AbstractMCMC ~/.julia/packages/AbstractMCMC/FSyVk/src/sample.jl:142
 [233] sample(rng::Xoshiro, model::DynamicPPL.Model{…}, sampler::DynamicPPL.Sampler{…}, N::Int64; chain_type::Type, resume_from::Nothing, initial_state::Nothing, progress::Bool, nadapts::Int64, discard_adapt::Bool, discard_initial::Int64, kwargs::@Kwargs{})
     @ Turing.Inference ~/.julia/packages/Turing/oFGEb/src/mcmc/hmc.jl:119
 [234] sample
     @ ~/.julia/packages/Turing/oFGEb/src/mcmc/hmc.jl:88 [inlined]
 [235] #sample#6
     @ ~/.julia/packages/Turing/oFGEb/src/mcmc/Inference.jl:321 [inlined]
 [236] sample(rng::Xoshiro, model::DynamicPPL.Model{typeof(foo), (:y,), (), (), Tuple{Vector{Float64}}, Tuple{}, DynamicPPL.DefaultContext}, alg::NUTS{AutoTracker, (), AdvancedHMC.DiagEuclideanMetric}, N::Int64)
     @ Turing.Inference ~/.julia/packages/Turing/oFGEb/src/mcmc/Inference.jl:312
Some type information was truncated. Use `show(err)` to see complete types.
```

</details>

Output from `] st`:
```
Status `~/tracker_test/Project.toml`
  [47edcb42] ADTypes v1.13.0
  [31c24e10] Distributions v0.25.117
  [da2b9cff] Mooncake v0.4.95
  [c3e4b0f8] Pluto v0.20.4
  [f3b207a7] StatsPlots v0.15.7
  [9f7883ad] Tracker v0.2.37
  [fce5fe82] Turing v0.36.2
```

**Reply by patrickm663** [reply]
Rerunning the above with Turing v0.34.1 works:

```
using Turing, ADTypes, Distributions, StatsPlots, Random, Tracker

@model function foo(y)
	τ ~ truncated(Normal(0, 1); lower=1e-5)
	return y ~ Normal(0, τ^2)
end

y = rand(Xoshiro(1), Normal(0, 2^2), 100);

ad = ADTypes.AutoTracker()

ch_tracker = sample(Xoshiro(2), foo(y), NUTS(; adtype=ad), 500)
```
Output:
```
┌ Info: Found initial step size
└   ϵ = 0.000390625
Sampling 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| Time: 0:00:01
Chains MCMC chain (500×13×1 Array{Float64, 3}):

Iterations        = 251:1:750
Number of chains  = 1
Samples per chain = 500
Wall duration     = 1.78 seconds
Compute duration  = 1.78 seconds
parameters        = τ
internals         = lp, n_steps, is_accept, acceptance_rate, log_density, hamiltonian_energy, hamiltonian_energy_error, max_hamiltonian_energy_error, tree_depth, numerical_error, step_size, nom_step_size

Summary Statistics
  parameters      mean       std      mcse   ess_bulk   ess_tail      rhat   ess_per_sec 
      Symbol   Float64   Float64   Float64    Float64    Float64   Float64       Float64 

           τ    1.9842    0.0710    0.0046   238.5636   414.1465    1.0243      134.1753

Quantiles
  parameters      2.5%     25.0%     50.0%     75.0%     97.5% 
      Symbol   Float64   Float64   Float64   Float64   Float64 

           τ    1.8527    1.9330    1.9836    2.0299    2.1220
```

Output from `] st` (had to also remove `Mooncake.jl` when downgrading Turing):
```
Status `~/tracker_test/Project.toml`
  [47edcb42] ADTypes v1.13.0
  [31c24e10] Distributions v0.25.117
  [c3e4b0f8] Pluto v0.20.4
  [f3b207a7] StatsPlots v0.15.7
  [9f7883ad] Tracker v0.2.37
⌃ [fce5fe82] Turing v0.34.1
Info Packages marked with ⌃ have new versions available and may be upgradable.
```
```

**Reply by patrickm663** [reply]
A more detailed MWE of a BNN comparing `Mooncake.jl` to `Tracker.jl`:

**TLDR:**
On a toy example with NNs of size ~70 and ~170, `Mooncake.jl` outperforms `Tracker.jl` :) However, TTFX when sampling from a BNN with Mooncake can be quite severe (e.g. 23 sec for 1 sample, vs 172 sec for 1 000 samples post-TTFX). There may be other bits like improvements in `Turing.jl` and `Lux.jl` coming through when Turing isn't pinned to v0.34.1. TTFX for Mooncake grows worse still when the size of the dataset increases -- for instance, 129 sec for 1 'warm-up' sample on a dataset of size `10x1000` vs 500 sec for 1 000 samples post-TTFX.

```
using Turing, ADTypes, Distributions, StatsPlots, Random, LinearAlgebra, Lux, ComponentArrays, Mooncake

# Toy problem
X = rand(Xoshiro(1), 10, 100);
y = sum(X; dims=1)/10

bnn_arch = Chain(Dense(10 => 4, tanh), Dense(4 => 4, tanh), Dense(4 => 1)) # 69 parameters

vector_to_parameters(ps_new, ps_) = ComponentArray(ps_new, getaxes(ps_))

function BNN(BNN_arch, N_samples, X_train, y_train, ad)
     ps_BNN, st_BNN = Lux.setup(Xoshiro(12345), BNN_arch) |> f64
     ps_BNN = ps_BNN |> ComponentArray

     @model function bayes_nn(xs, ys, BNN_arch, ps_BNN, sig, n_params, ::Type{T} = Float64) where {T}
     
             # Priors
             ## Sample the parameters
             σ ~ truncated(Normal(0, 1); lower=1e-9)
             parameters ~ MvNormal(zeros(n_params), (sig^2) .* I)
     
             ## Forward NN to make predictions
             μ, st_BNN = Lux.apply(BNN_arch, xs, vector_to_parameters(parameters, ps_BNN), st_BNN)
     
             # Likelihood
             ys ~ MvNormal(vec(μ), (σ^2) .* I)

             return nothing
     end

     n_params = Lux.parameterlength(BNN_arch)

     alpha = 0.9
     sig = sqrt(1.0 / alpha)

     BNN_inference = bayes_nn(X_train |> f64, vec(y_train) |> f64, BNN_arch, ps_BNN, sig, n_params)

     chains = sample(Xoshiro(22), BNN_inference, NUTS(; adtype=ad), N_samples; discard_adapt=false)

     return chains, ps_BNN, st_BNN
end

ad = ADTypes.AutoMooncake(; config=nothing)

chains, ps_BNN, st_BNN = BNN(bnn_arch, 1_000, X, y, ad);
```

Run-time:
`Mooncake.jl` (Turing v0.36.2)
```
Chains MCMC chain (1000×82×1 Array{Float64, 3}):

Iterations        = 1:1:1000
Number of chains  = 1
Samples per chain = 1000
Wall duration     = 94.86 seconds
Compute duration  = 94.86 seconds
```
<details>
<summary> `] st` </summary>

```
Status `~/tracker_test/Project.toml`
  [47edcb42] ADTypes v1.13.0
  [b0b7db55] ComponentArrays v0.15.24
  [31c24e10] Distributions v0.25.117
  [b2108857] Lux v1.7.0
  [da2b9cff] Mooncake v0.4.95
  [c3e4b0f8] Pluto v0.20.4
  [f3b207a7] StatsPlots v0.15.7
  [9f7883ad] Tracker v0.2.37
  [fce5fe82] Turing v0.36.2
```

</details>

`Tracker.jl` (Turing v0.34.1)
```
Chains MCMC chain (1000×82×1 Array{Float64, 3}):

Iterations        = 1:1:1000
Number of chains  = 1
Samples per chain = 1000
Wall duration     = 136.57 seconds
Compute duration  = 136.57 seconds
```
<details>
<summary> `] st` </summary>

```
Status `~/tracker_test/Project.toml`
  [47edcb42] ADTypes v1.13.0
  [b0b7db55] ComponentArrays v0.15.24
  [31c24e10] Distributions v0.25.117
⌃ [b2108857] Lux v1.0.5
  [c3e4b0f8] Pluto v0.20.4
  [f3b207a7] StatsPlots v0.15.7
  [9f7883ad] Tracker v0.2.37
⌃ [fce5fe82] Turing v0.34.1
```

</details>

In addition, increasing the size of the NN still puts `Mooncake.jl` ahead:
```
bnn_arch = Chain(Dense(10 => 8, tanh), Dense(8 => 8, tanh), Dense(8 => 1)) # 169 parameters
```

`Mooncake.jl` (Turing v0.36.2)
```
Chains MCMC chain (1000×182×1 Array{Float64, 3}):

Iterations        = 1:1:1000
Number of chains  = 1
Samples per chain = 1000
Wall duration     = 171.99 seconds
Compute duration  = 171.99 seconds

```

`Tracker.jl` (Turing v0.34.1)
```
Chains MCMC chain (1000×182×1 Array{Float64, 3}):

Iterations        = 1:1:1000
Number of chains  = 1
Samples per chain = 1000
Wall duration     = 228.82 seconds
Compute duration  = 228.82 seconds
```

**Reply by willtebbutt** [reply]
It's very odd that the TTFX grows as the dataset size increases (it should be quite consistent, and that's typically been the experience I and other have had).

Would you mind seeing what happens if you improve the type stability of the above code snippet by declaring `y` and `X` to be `const`?

**Reply by patrickm663** [reply]
Sure, will try. On a fresh session, it looks like it just hangs when running on one sample until it eventually gets going EDIT: I mean for it to print that it has found an initial step size

**Reply by willtebbutt** [reply]
Oh, wait, scratch that. You're passing them in directly, so my point makes no sense. Hmmm... I'll have a go locally and see if I can figure out what might be happening.

**Reply by willtebbutt** [reply]
Okay, so it looks like it really is all in the compile time, as you say. When I generate benchmarks for the above using TuringBenchmarking.jl, I see something along the lines of
```julia
2-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "evaluation" => 2-element BenchmarkTools.BenchmarkGroup:
	  tags: []
	  "linked" => Trial(11.629 μs)
	  "standard" => Trial(11.511 μs)
  "gradient" => 1-element BenchmarkTools.BenchmarkGroup:
	  tags: []
	  "AutoMooncake{Nothing}(nothing)" => 2-element BenchmarkTools.BenchmarkGroup:
		  tags: ["AutoMooncake{Nothing}(nothing)"]
		  "linked" => Trial(68.880 μs)
		  "standard" => Trial(65.467 μs)
```
So the gradient / primal ratio is about 6 -- not inspirational, but also not atrocious. Having also taken a look at some profiling, it doesn't look like type instability problems.

However, we are generating absolutely _enormous_ types for some Turing.jl models, and it looks like this one in particular is potentially causing trouble. I'll have to have a think about how we might tackle this.

edit: to be clear, I do generally expect Mooncake to have slow-ish compile times when compared to something like ForwardDiff or Tracker, but I wouldn't typically expect them to be as large as we're seeing here.

**Reply by patrickm663** [reply]
Thanks for looking into this @willtebbutt
